src/adapter/src/catalog.rs

```diff
@@ -1487,6 +1487,23 @@ impl Source {
             DataSourceDesc::Source | DataSourceDesc::Introspection(_) => false,
         }
     }
+
+    /// Type of the source.
+    pub fn source_type(&self) -> &str {
+        match &self.data_source {
+            DataSourceDesc::Ingestion(ingestion) => ingestion.desc.name(),
+            DataSourceDesc::Source => "subsource",
+            DataSourceDesc::Introspection(_) => "source",
```

Comment:
This is the only block I don't feel like I can review because I don't know what's going on. Adding more tests would make me feel better about this.